graders/python/python_autograder/code_feedback.py

```diff
@@ -299,6 +299,122 @@ def bad(msg):
+        - ``accuracy_critical``: If true, grading will halt on failure.
+        - ``report_failure``: If true, feedback will be given on failure.
+        """
+
+        def bad(msg):
+            if report_failure:
+                cls.add_feedback(msg)
+            if accuracy_critical:
+                cls.finish("")
+            else:
+                return False
+
+        if data is None:
+            return bad(f"{name} is None or not defined")
+
+        if not isinstance(data, dict):
+            return bad(f"{name} is not a dict")
+
+        if value_type is not None:
+            for value in data.values():
+                if not isinstance(value, value_type):
+                    return bad(
+                        f"{name} has the wrong type for value {value}, expected type {value_type}"
+                    )
+        if key_type is not None:
+            for key in data.keys():
+                if not isinstance(key, key_type):
+                    return bad(
+                        f"{name} has the wrong type for key {key}, expected type {key_type}"
+                    )
+
+        if partial_keys is None:
+            if check_only_values:
+                if len(ref.values()) != len(data.values()):
+                    return bad(
+                        f"{name} has the wrong number of values: expected {len(ref.values())}, got {len(data.values())}"
+                    )
+
+            if len(ref) != len(data):  # this is default length of keys check
+                return bad(
+                    f"{name} has the wrong number of entries: expected {len(ref)}, got {len(data)}"
+                )
+        check_partial_keys = partial_keys is not None and len(partial_keys) >= 1
+
+        if check_only_keys or check_only_values or check_partial_keys:
+            
+            partial_keys_valid = False
+            if partial_keys is not None and len(partial_keys) >= 1:
+                for partial_key in partial_keys:
+                    if partial_key not in data:
+                        return bad(f"{name} does not contain key {partial_key}")
+                partial_keys_valid =  True
+
+            check_keys = False
+            if check_only_keys:
+                for key in data.keys():
+                    if key not in ref.keys():
+                        return bad(f"{name} contains an extra key: {key}")
+                check_keys = True
+
+            check_values = False
+            if check_only_values:
+                if len(ref.values()) != len(data.values()):
+                    return f"{name} has the wrong length for values: expected {len(ref.values())}, got {len(data.values())}"
+                for value in data.values():
+                    if value not in ref.values():
+                        return bad(f"{name} contains an extra value: {value}")
+                check_values = True
+
+            if check_only_keys and check_only_values and check_partial_keys:
```

Comment:
Is there a simpler way to write the if-cases of this function below? There should be a way of doing that avoids checking every single case.

Reply:
I updated this in the latest push